packages/oauth-client/README.md

```diff
@@ -0,0 +1,254 @@
+
+**Kind**: global class  
+<a name="EarthdataLoginClient+getTokenUsername"></a>
+
+### earthdataLoginClient.getTokenUsername(params) ⇒ <code>Promise.&lt;string&gt;</code>
+Query the Earthdata Login API for the UID associated with a token
+
+**Kind**: instance method of [<code>EarthdataLoginClient</code>](#EarthdataLoginClient)  
+**Returns**: <code>Promise.&lt;string&gt;</code> - the UID associated with the token  
+
+| Param | Type | Description |
+| --- | --- | --- |
+| params | <code>Object</code> |  |
+| params.onBehalfOf | <code>string</code> | the Earthdata Login client id of the   app requesting the username |
+| params.token | <code>string</code> | the Earthdata Login token |
+| [params.xRequestId] | <code>string</code> | a string to help identify the request   in the Earthdata Login logs |
+
+<a name="OAuthClient"></a>
+
+## OAuthClient
+A generic authorization client
+
+**Kind**: global class  
+
+* [OAuthClient](#OAuthClient)
+    * [new OAuthClient(params)](#new_OAuthClient_new)
+    * [.getAuthorizationUrl([state])](#OAuthClient+getAuthorizationUrl) ⇒ <code>string</code>
+    * [.getAccessToken(authorizationCode)](#OAuthClient+getAccessToken) ⇒ <code>Promise.&lt;Object&gt;</code>
+    * [.postRequest(params)](#OAuthClient+postRequest) ⇒ <code>CancelableRequest.&lt;Response.&lt;unknown&gt;&gt;</code>
+    * [.getRequest(params)](#OAuthClient+getRequest) ⇒ <code>CancelableRequest.&lt;Response.&lt;unknown&gt;&gt;</code>
+    * [.refreshAccessToken(refreshToken)](#OAuthClient+refreshAccessToken) ⇒ <code>Promise.&lt;Object&gt;</code>
+
+<a name="new_OAuthClient_new"></a>
+
+### new OAuthClient(params)
+
+| Param | Type | Description |
+| --- | --- | --- |
+| params | <code>Object</code> |  |
+| params.clientId | <code>string</code> | see example |
+| params.clientPassword | <code>string</code> | see example |
+| params.loginUrl | <code>string</code> | see example |
+| params.redirectUri | <code>string</code> | see example |
+
+**Example**  
+```js
+const oAuth2Provider = new OAuthClient({
+  clientId: 'my-client-id',
+  clientPassword: 'my-client-password',
+  loginUrl: 'https://earthdata.login.nasa.gov',
+  redirectUri: 'http://my-api.com'
+});
+```
+<a name="OAuthClient+getAuthorizationUrl"></a>
+
+### oAuthClient.getAuthorizationUrl([state]) ⇒ <code>string</code>
+Get a URL of the Login authorization endpoint
+
+**Kind**: instance method of [<code>OAuthClient</code>](#OAuthClient)  
+**Returns**: <code>string</code> - the Login authorization URL  
+
+| Param | Type | Description |
+| --- | --- | --- |
+| [state] | <code>string</code> | an optional state to pass to login Client |
+
+<a name="OAuthClient+getAccessToken"></a>
+
+### oAuthClient.getAccessToken(authorizationCode) ⇒ <code>Promise.&lt;Object&gt;</code>
+Given an authorization code, request an access token and associated
+information from the login service.
+
+Returns an object with the following properties:
+
+- accessToken
+- refreshToken
+- username
```

Comment:
For Cognito, the api doesn't return `username` or `endpoint`(where the username is extracted from)

Reply:
Huh. Curious that I wasn't seeing failures there. That should blow up if `endpoint` is undefined. I was testing Cognito from the unit tests against the actual API but maybe I had some response handling in there that masked this.

In any case, here's a fix to make `username` undefined if `endpoint` isn't available. I could also remove `username` from the response in this case but, because this is on the parent `OAuthClient` I was thinking it might be best to not do that in case EarthdataLogin ever returns an invalid endpoint. I think it'd be confusing in that case to just not have any hint of `username`. Open to push-back though.

https://github.com/nasa/cumulus/pull/2209/commits/1af43af0ec6b4689d27f1d6bcdc5ab37d4fd83fb